test/taskrun_test.go

```diff
@@ -0,0 +1,61 @@
+// +build e2e
+
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package test
+
+import (
+	"testing"
+
+	knativetest "github.com/knative/pkg/test"
+	tb "github.com/tektoncd/pipeline/test/builder"
+)
+
+func TestTaskRunFailure(t *testing.T) {
+	c, namespace := setup(t)
+	t.Parallel()
+
+	knativetest.CleanupOnInterrupt(func() { tearDown(t, c, namespace) }, t.Logf)
+	defer tearDown(t, c, namespace)
+
+	t.Logf("Creating Task and TaskRun in namespace %s", namespace)
+	task := tb.Task("failing-task", namespace, tb.TaskSpec(
+		tb.Step("hello", "busybox",
+			tb.Command("/bin/sh"), tb.Args("-c", "echo hello"),
+		),
+		tb.Step("exit", "busybox",
+			tb.Command("/bin/sh"), tb.Args("-c", "exit 1"),
+		),
+		tb.Step("world", "busybox",
+			tb.Command("/bin/sh"), tb.Args("-c", "sleep 30s"),
+		),
+	))
+	if _, err := c.TaskClient.Create(task); err != nil {
+		t.Fatalf("Failed to create Task: %s", err)
+	}
+	taskRun := tb.TaskRun("failing-taskrun", namespace, tb.TaskRunSpec(
+		tb.TaskRunTaskRef("failing-task"),
+	))
+	if _, err := c.TaskRunClient.Create(taskRun); err != nil {
+		t.Fatalf("Failed to create TaskRun: %s", err)
+	}
+
+	t.Logf("Waiting for TaskRun in namespace %s to fail", namespace)
+	if err := WaitForTaskRunState(c, "failing-taskrun", TaskRunFailed("failing-taskrun"), "TaskRunFailed"); err != nil {
+		t.Errorf("Error waiting for TaskRun to finish: %s", err)
```

Comment:
would be nice to check the status of the `taskrun` after it failed and make sure step 2 has failure and step 3 didn't run. Before this fix, it was failing eventually by waiting for the default timeout, somehow if we can check that's not the failure reason.

Reply:
Right, good point, I should validate that all steps after the `exit 1` are terminated and in the same error state :angel:
That raise a question : in the case of steps being "skipped" because a previous one failed, should we return another exit code than the default `1` ? (to be able to detect such cases)

Reply:
If we do that, do we end up with confusion if a step exits with a non-0/1 exit code on purpose? I'd lean towards using `Reason` or `Message` rather than `ExitCode`.

Reply:
:thinking: it won't change the status of the `TaskRun`. It would only be useful to detect that some of the container didn't have to execute. We could also "copy" the exit code of the failed process to the later ones.

Reply:
not sure what the best option is, would be nice to know it didn't execute, anything here makes sense? http://tldp.org/LDP/abs/html/exitcodes.html

Reply:
We discussed quiclky with @abayer on slack : 

- `exit 0` on skipped containers (the one after the failure)
- updating the TaskRun status to point to the failed step
- updating the steps status with some skipped messages for the steps that were skip

Reply:
- exit 0 on skipped containers (the one after the failure) 👍 
 - updating the TaskRun status to point to the failed step 🤔 but is that an api change to add that? It already has an array of steps status which should contain the same info 
 - updating the steps status with some skipped messages for the steps that were skip 🤔 I wonder if we are over thinking this. If a step fails, that should be clear enough why the taskrun failed

Reply:
Ok, I'll go for exit code 0 in here (and make sure the TaskRun status references the correct step as a failure) — we have time to think about skip messages & co later on, in a follow-up.